### PR TITLE
detect mtu from multiple sources

### PR DESCRIPTION
**Pull request by ozdanborne**
## Description

Check all four sources of MTU and, if set, make sure they match.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/migration/convert/mtu.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			install.Spec.CalicoNetwork = &operatorv1.CalicoNetworkSpec{}
+		}
+		install.Spec.CalicoNetwork.MTU = detectedMTU
+	}
+
+	return nil
+}
+
+// compareMTU compares two MTU values and ensures they match if both are set
+func compareMTU(m *string, detectedMTU *int32) (*int32, error) {
```
Comment (ozdanborne) [inline pkg/controller/migration/convert/mtu.go:81]: @tmjd I wrote this code and even I find it a little hard to follow. Any tips on how you would have done it?

File: pkg/controller/migration/convert/mtu.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	m, err := c.node.getEnv(ctx, c.client, "calico-node", "FELIX_IPINIPMTU")
+	if err != nil {
+		return err
+	}
+	detectedMTU, err = compareMTU(m, detectedMTU)
+	if err != nil {
+		return err
+	}
+
+	m, err = c.node.getEnv(ctx, c.client, "calico-node", "FELIX_VXLANMTU")
```
Comment (tmjd) [inline pkg/controller/migration/convert/mtu.go:22]: Isn't there a constant for `calico-node`?

File: pkg/controller/migration/convert/mtu.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+			if mtu == nil {
+				// if not set, install-cni will use a known default mtu of 1500
+				mtu = new(int32)
+				*mtu = 1500
+			}
+
+			// compare against current mtu.
+			if curMTU != nil && *curMTU != *mtu {
+				return fmt.Errorf("detected mtu %s=%d does not match detected mtu %s=%d", "CNI_MTU", *mtu, curMTUSrc, *curMTU)
```
Comment (tmjd) [inline pkg/controller/migration/convert/mtu.go:52]: I think this should be an ErrIncompatibleCluster error.

File: pkg/controller/migration/convert/mtu.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+			}
+			curMTU, curMTUSrc = mtu, "CNI_MTU"
+
+		} else {
+			// user must have hardcoded their CNI instead of using the cni templating engine.
+			// use the hardcoded value.
+			mtu := int32(c.calicoCNIConfig.MTU)
+			if curMTU != nil && *curMTU != mtu {
+				return ErrIncompatibleCluster{"MTUs across IPIP, VXLAN, and Wireguard must match."}
```
Comment (tmjd) [inline pkg/controller/migration/convert/mtu.go:61]: Should include the 'curMTUSrc' and value in this error message.

File: pkg/controller/migration/convert/mtu.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	m, err := c.node.getEnv(ctx, c.client, container, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if m == nil {
+		return nil, nil
+	}
+
+	// if it's not nil, convert it to a number, and compare it against the currently detected mtu
```
Comment (tmjd) [inline pkg/controller/migration/convert/mtu.go:89]: This comment isn't correct anymore.

File: pkg/controller/migration/convert/mtu.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	i, err := strconv.ParseInt(*m, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't convert %s to integer: %v", *m, err)
```
Comment (tmjd) [inline pkg/controller/migration/convert/mtu.go:92]: Should we include the container/key in this error message?

Comment (ozdanborne) [reply pkg/controller/migration/convert/mtu.go:95]: The calling function already wraps the error with that information. I think either place works fine, no?
